examples/lox/src/language/type-system/lox-type-checking.ts

```diff
@@ -0,0 +1,287 @@
+                 * - additionally, validations for the assigned values to the expected parameter( type)s are derived */
+                inferenceRuleForCalls: {
+                    filter: isMemberCall,
+                    matching: (domainElement: MemberCall) => isFunctionDeclaration(domainElement.element?.ref) && domainElement.element!.ref.name === functionName,
+                    inputArguments: (domainElement: MemberCall) => domainElement.arguments
+                },
+            });
+        }
+
+        // TODO support lambda (type references)!
+
+        /**
+         * TODO Delayed:
+         * - (classType: Type) => Type(for output)
+         * - WANN werden sie aufgelöst? bei erster Verwendung?
+         * - WO wird das verwaltet? im Kind? im Type? im TypeGraph?
+         */
+
+        // class types (nominal typing):
+        if (isClass(node)) {
+            const className = node.name;
+            classKind.createClassType({
+                className,
+                superClasses: node.superClass?.ref ? classKind.getClassType(node.superClass.ref.name) : undefined, // TODO delayed
+                fields: node.members
+                    .filter(m => isFieldMember(m)).map(f => f as FieldMember) // only Fields, no Methods
+                    .map(f => <NameTypePair>{
+                        name: f.name,
+                        type: mapType(f.type), // TODO delayed
+                    }),
+                // inference rule for declaration
+                inferenceRuleForDeclaration: (domainElement: unknown) => domainElement === node,
+                // inference ruleS(?) for objects/class literals conforming to the current class
+                inferenceRuleForLiteral: { // <InferClassLiteral<MemberCall>>
+                    filter: isMemberCall,
+                    matching: (domainElement: MemberCall) => isClass(domainElement.element?.ref) && domainElement.element!.ref.name === className,
+                    inputValuesForFields: (_domainElement: MemberCall) => new Map(), // values for fields don't matter for nominal typing
+                },
+                inferenceRuleForReference: { // <InferClassLiteral<TypeReference>>
+                    filter: isTypeReference,
+                    matching: (domainElement: TypeReference) => isClass(domainElement.reference?.ref) && domainElement.reference!.ref.name === className,
+                    inputValuesForFields: (_domainElement: TypeReference) => new Map(), // values for fields don't matter for nominal typing
+                },
+                // inference rule for accessing fields
+                inferenceRuleForFieldAccess: (domainElement: unknown) => isMemberCall(domainElement) && isFieldMember(domainElement.element?.ref) && domainElement.element!.ref.$container === node
+                    ? domainElement.element!.ref.name : 'N/A', // as an alternative, use 'InferenceRuleNotApplicable' instead, what should we recommend?
```

Comment:
Very dense information, why do we need those inference rules?

Reply:
Since there are multiple AstNodes in LOX which are related to class types.

How would you structure these inference rules?